rfc/rfc-46/rfc-46.md

```diff
@@ -156,13 +187,76 @@ Following major components will be refactored:
 3. `HoodieRealtimeRecordReader`s 
    1. API will be returning opaque `HoodieRecord` instead of raw Avro payload
 
+### Config for Record Merge
+The MERGE_CLASS_NAME config is engine-aware. If you are not specified the MERGE_CLASS_NAME, MERGE_CLASS_NAME will be specified default according to your engine type.
+
+### Public Api in HoodieRecord
+Because we implement different types of records, we need to transfer some func in AvroUtils into HoodieRecord for different data(avro, InternalRow, RowData).
+Its public API will look like following:
+
+```java
+class HoodieRecord {
+
+   /**
+    * Get column in record to support RDDCustomColumnsSortPartitioner
+    */
+   Object getRecordColumnValues(Schema recordSchema, String[] columns, boolean consistentLogicalTimestampEnabled);
+
+   /**
+    * Support bootstrap.
+    */
+   HoodieRecord mergeWith(HoodieRecord other) throws IOException;
+
+   /**
+    * Rewrite record into new schema(add meta columns)
+    */
+   HoodieRecord rewriteRecord(Schema recordSchema, Properties props, Schema targetSchema) throws IOException;
+
+   /**
+    * Support schema evolution.
+    */
+   HoodieRecord rewriteRecordWithNewSchema(Schema recordSchema, Properties props, Schema newSchema, Map<String, String> renameCols) throws IOException;
+
+   HoodieRecord addMetadataValues(Schema recordSchema, Properties props, Map<HoodieMetadataField, String> metadataValues) throws IOException;
+
+   /**
+    * Is deleted.
+    */
+   boolean isPresent(Schema recordSchema, Properties props) throws IOException;
+
+   /**
+    * Is EmptyRecord. Generated by ExpressionPayload.
+    */
+   boolean shouldIgnore(Schema recordSchema, Properties props) throws IOException;
+
+   /**
+    * This method used to extract HoodieKey not through keyGenerator.
+    */
+   HoodieRecord expansion(
+           String payloadClass,
+           String preCombineField,
+           Option<Pair<String, String>> simpleKeyGenFieldsOpt,
+           Boolean withOperation,
+           Option<String> partitionNameOp,
+           Boolean populateMetaFieldsOp) throws IOException;
+
+   /**
+    * This method used to extract HoodieKey through keyGenerator. This method used in ClusteringExecutionStrategy.
+    */
+   HoodieRecord transform(Properties props, Option<BaseKeyGenerator> keyGen);
+
+   Option<IndexedRecord> toIndexedRecord(Schema schema, Properties props) throws IOException;
```

Comment:
What's the purpose of this method? Are we using it to fallback to Avro? 

If that's the case then we should:

1. Return `HoodieAvroRecord` instead of `IndexedRecord` 
2. Make sure we annotate it as the one which usages should be very restricted (mostly for the time we migrate and in a few exceptions)

Reply:
fixed